content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,530 @@
+mark(
+  dplyr = bind_rows(mtcars_tbl, mtcars_tbl),
+  vctrs = vec_rbind(mtcars_tbl, mtcars_tbl)
+) %>%
+  select(expression, median)
+```
+
+As for column-binding:
+
+```{r bm-bind-cols}
+tbl <- tibble(year = rep(1974L, nrow(mtcars_tbl)))
+
+mark(
+  dplyr = bind_cols(mtcars_tbl, tbl),
+  vctrs = vec_cbind(mtcars_tbl, tbl)
+) %>%
+  select(expression, median)
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/tune#636](https://github.com/tidymodels/tune/pull/636).
+
+#### `group_by()` + `summarize()`
+
+In general, the introduction of groups makes these substitutions much trickier. In those cases, it's likely best to weigh (via profiling) how significant the slowdown is and, if it's not too bad, opt not to make any changes. For code that relies on `group_by()` and sees heavy traffic, see `vctrs::list_unchop()`, `vctrs::vec_chop()`, and `vctrs::vec_rep_each()`.
+
+### Tibbles
+
+Tibbles are great, and I don't want to interface with any other data frame-y thing. Some notes:
+
+- `as_tibble()` on a tibble is not "free":
+
+```{r bm-as-tibble}
+mark(
+  on_tbl_df = as_tibble(mtcars_tbl),
+  on_data.frame = as_tibble(mtcars, rownames = "make_model")
+) %>% 
+  select(expression, median)
+```
+
+- Building a tibble from scratch using `tibble()` actually takes quite a while as well. `tibble()` handles vector recycling and name checking, builds columns sequentially, all that good stuff. If you need that, use `tibble()`, but if you're building a tibble from well-understood inputs, use `new_tibble(list())`. For a middle ground between `tibble()` and `new_tibble(list())` in terms of both performance and safety, use the `df_list()` function from the vctrs package in place of `list()`.
+
+```{r bm-tibble}
+mark(
+  tibble = tibble(a = 1:2, b = 3:4),
+  new_tibble_df_list = new_tibble(df_list(a = 1:2, b = 3:4), nrow = 2),
+  new_tibble_list = new_tibble(list(a = 1:2, b = 3:4), nrow = 2)
+) %>% 
+  select(expression, median)
+```
+
+Note that `new_tibble()` *will not check the lengths of its inputs.* Carry out simple recycling yourself, and be sure to use the `nrow` argument to get basic length checks.
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#945](https://github.com/tidymodels/parsnip/pull/932), [tidymodels/parsnip#934](https://github.com/tidymodels/parsnip/pull/934), [tidymodels/parsnip#929](https://github.com/tidymodels/parsnip/pull/929), [tidymodels/parsnip#923](https://github.com/tidymodels/parsnip/pull/923), [tidymodels/parsnip#902](https://github.com/tidymodels/parsnip/pull/902), [tidymodels/dials#277](https://github.com/tidymodels/dials/pull/277), and [tidymodels/tune#637](https://github.com/tidymodels/tune/pull/637).
+
+### Becoming Join-Critical
+
+Two truths:
+
+- dplyr joins are a remarkably safe and powerful way to synthesize data sources.
+
+- One ought to ask themselves "does this really need to be a join?" when combining data sources in package code.
```

Comment:
Another places I see joins where another approach would be better is lookup tables. Character matching as described [here](https://adv-r.hadley.nz/subsetting.html?q=lookup#lookup-tables) is a great alternative. This example is a bit contrived but illustrates the point:

``` r
library(tidyverse)
library(bench)

mtcars_tbl <- as_tibble(mtcars, rownames = "make_model")

add_size <- function() {
  lu <- c("4" = "small", "6" = "medium", "8" = "big")
  mtcars_tbl$size <- unname(lu[as.character(mtcars_tbl$cyl)])
  mtcars_tbl
}

lu_df <- tibble(cyl = c(4,6,8), size = c("small", "medium", "big"))

mark(
  join = left_join(mtcars_tbl, lu_df, by = "cyl"),
  lookup = add_size()
) |> 
  select(expression, median)
#> # A tibble: 2 × 2
#>   expression   median
#>   <bch:expr> <bch:tm>
#> 1 join        451.9µs
#> 2 lookup       25.9µs
```

<sup>Created on 2023-04-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>